Qq/Macro.lean

```diff
@@ -414,6 +414,41 @@ partial def quoteExpr : Expr → QuoteM Expr
   | .proj n i e => return mkApp3 (.const ``Expr.proj []) (toExpr n) (toExpr i) (← quoteExpr e)
   | .mdata _ e => quoteExpr e
 
+/--
+Translates an arbitrary local context to a context of
+Q-annotated expressions. Used by `by_elabq` and `run_tacq`.
+-/
+def quoteLCtx (ctx : LocalContext) (levelNames : List Name) :
+    UnquoteM (LocalContext × Array Expr) := do
+  let mut quotedCtx := LocalContext.empty
+  let mut assignments : Array Expr := #[]
+  for nm in levelNames do
+    let fid ← mkFreshFVarId
+    quotedCtx := quotedCtx.mkLocalDecl
+      fid nm (mkConst ``Level) .default .default
+    modify fun s => { s with levelBackSubst := s.levelBackSubst.insert (.param nm) (.fvar fid) }
+    assignments := assignments.push (toExpr (Level.param nm))
+  for decl in ctx do
+    if decl.kind != .default then continue
+    let fid ← mkFreshFVarId
+    let type ← instantiateMVars decl.type
+    let quotedType ← quoteExpr type
+    quotedCtx := quotedCtx.mkLocalDecl fid decl.userName
+      (mkApp (mkConst ``Quoted) quotedType) decl.binderInfo
+    assignments := assignments.push (toExpr (Expr.fvar decl.fvarId))
+    if decl.isLet then
+      let eqFid ← mkFreshFVarId
+      let level ← getLevel type
+      let quotedLevel ← (quoteLevel (← instantiateLevelMVars level))
+      let quotedValue ← (quoteExpr (← instantiateMVars decl.value))
+      quotedCtx := quotedCtx.mkLocalDecl eqFid (← mkFreshUserName `eq)
```

Comment:
Perhaps we should use ``decl.userName ++ `eq`` as the name here?

Reply:
done